daemon/src/setup_contract.rs

```diff
@@ -221,12 +226,291 @@ pub async fn new(
         identity: sk,
         identity_counterparty: params.other.identity_pk,
         revocation: rev_sk,
+        revocation_pk_counterparty: other_punish.revocation_pk,
         publish: publish_sk,
-        address: params.own.address,
+        publish_pk_counterparty: other_punish.publish_pk,
+        maker_address: params.maker().address.clone(),
+        taker_address: params.taker().address.clone(),
         lock: (signed_lock_tx.extract_tx(), lock_desc),
         commit: (commit_tx, msg1.commit, commit_desc),
         cets,
         refund: (refund_tx, msg1.refund),
+        maker_lock_amount: params.maker().lock_amount,
+        taker_lock_amount: params.taker().lock_amount,
+        revoked_commit: Vec::new(),
+    })
+}
+
+pub async fn roll_over(
+    mut sink: impl Sink<RollOverMsg, Error = anyhow::Error> + Unpin,
+    mut stream: impl FusedStream<Item = RollOverMsg> + Unpin,
+    (oracle_pk, announcement): (schnorrsig::PublicKey, oracle::Announcement),
+    cfd: Cfd,
+    our_role: Role,
+    dlc: Dlc,
+) -> Result<Dlc> {
+    let sk = dlc.identity;
+    let pk = PublicKey::new(secp256k1_zkp::PublicKey::from_secret_key(SECP256K1, &sk));
+
+    let (rev_sk, rev_pk) = crate::keypair::new(&mut rand::thread_rng());
+    let (publish_sk, publish_pk) = crate::keypair::new(&mut rand::thread_rng());
+
+    let own_punish = PunishParams {
+        revocation_pk: rev_pk,
+        publish_pk,
+    };
+
+    sink.send(RollOverMsg::Msg0(RollOverMsg0 {
+        revocation_pk: rev_pk,
+        publish_pk,
+    }))
+    .await
+    .context("Failed to send Msg0")?;
+    let msg0 = stream
+        .select_next_some()
+        .await
+        .try_into_msg0()
+        .context("Failed to read Msg0")?;
+
+    let maker_lock_amount = dlc.maker_lock_amount;
+    let taker_lock_amount = dlc.taker_lock_amount;
+    let payouts = HashMap::from_iter([(
+        // TODO : we want to support multiple announcements
+        Announcement {
+            id: announcement.id.0,
+            nonce_pks: announcement.nonce_pks.clone(),
+        },
+        payout_curve::calculate(
+            cfd.order.price,
+            cfd.quantity_usd,
+            maker_lock_amount,
+            (taker_lock_amount, cfd.order.leverage),
+        )?,
+    )]);
+
+    // unsign lock tx because PartiallySignedTransaction needs an unsigned tx
+    let unsigned_lock_tx = Transaction {
+        version: 0,
+        lock_time: dlc.lock.0.lock_time,
+        input: dlc
+            .lock
+            .0
+            .input
+            .iter()
+            .map(|txin| TxIn {
+                previous_output: txin.previous_output,
+                script_sig: txin.script_sig.clone(),
+                sequence: txin.sequence,
+                witness: vec![],
+            })
+            .collect(),
+        output: dlc.lock.0.output.clone(),
+    };
```

Comment:
I think this would be easier to understand if you would make a copy of `dlc.lock` and iterate over the inputs to clear the `witness`.